File: Android App/app/src/main/java/revert/common/ColorModule.java
package revert.common;

import android.graphics.Color;

import revert.shredsheets.models.SessionModel;
import revert.shredsheets.models.themes.ShredsheetsTheme;

public class ColorModule {

    public static int getHsvContrastingColor(int color) {
        float[] hsv = new float[3]; //{ h = 0..360, s = 0..1, v = 0..1 }
        Color.colorToHSV(color, hsv);
        float saturation = hsv[1];
        float value = hsv[2];
        float combined = saturation * value;
        int rgbValues = (int) (255 * combined);
        return Color.rgb(rgbValues, rgbValues, rgbValues);
    }

    public static int getContrastingColor(int color) {
        return Color.rgb(255 - Color.red(color), 255 - Color.green(color), 255 - Color.blue(color));
    }

    public static int getMaximallyContrastingColor(int color) {
        double y = (double)(299 * Color.red(color) + 587 * Color.green(color) + 114 * Color.blue(color)) / (double)1000;
        return y >= 128 ? Color.parseColor("#333333") : Color.parseColor("#F0F0F0");
    }

    public static int getModeratelyContrastingColor(int color) {
        double y = (double)(299 * Color.red(color) + 587 * Color.green(color) + 114 * Color.blue(color)) / (double)1000;
        return y >= 128 ? Color.parseColor("#555555") : Color.parseColor("#CCCCCC");
    }

    public static int getMinimallyContrastingColor(int color) {
        double y = (double)(299 * Color.red(color) + 587 * Color.green(color) + 114 * Color.blue(color)) / (double)1000;
        return y >= 128 ? Color.parseColor("#888888") : Color.parseColor("#666666");
    }


    public static int getIntervalColor(String intervalName) {
        SessionModel session = SessionModel.getInstance();
        ShredsheetsTheme theme = session.getTheme();

        int[] intervalColors = theme.getIntervalColors();

        switch (intervalName) {
            case "h":
                return intervalColors[0];
            case "W":
                return intervalColors[1];
            case "W+h":
                return intervalColors[2];
            default:
                return intervalColors[2];
        }
    }
}

File: Shredsheets 3/core/src/com/bm/shredsheets/ColorModule.kt
package com.bm.shredsheets

import com.badlogic.gdx.graphics.Color
import com.bm.shredsheets.models.SessionModel
import com.bm.shredsheets.models.themes.ShredsheetsTheme

object ColorModule {

    var darkColor = Color(.15f, .15f, .15f, 1f)
    var lightColor = Color(.925f, .925f, .925f, 1f)

    fun getMaximallyContrastingColor(color: Color): Color {
        var hsv = FloatArray(3)
        color.toHsv(hsv)

        if (hsv[2] > .5)
            return darkColor
        else
            return lightColor
    }

    fun getModeratelyContrastingColor(color: Int): Color {
        var colorItem = Color(color)
        val y = 299 * colorItem.r + 587 * colorItem.g + 114 * colorItem.b.toDouble() / 1000.toDouble()
        return if (y >= 128) Color(.35f, .35f, .35f, 1f) else Color(.8f, .8f, .8f, 1f)
    }

    var minDark = Color(.4f, .4f, .4f, 1f)
    var minLight = Color(.6f, .6f, .6f, 1f)
    var midDark = Color(.3f, .3f, .3f, 1f)

    fun getMinimallyContrastingColor(color: Color): Color {
        var hsv = FloatArray(3)
        color.toHsv(hsv)

        if (hsv[2] <= .4f)
            return minLight
        else if (hsv[2] >= .6f)
            return minDark
        else
            return midDark
    }

    fun getIntervalColor(intervalName: String?): Color {
        val session: SessionModel = SessionModel.instance
        val theme: ShredsheetsTheme = session.theme

        val intervalColors: Array<Color> = theme.intervalColors
        return when (intervalName) {
            "h" -> intervalColors[0]
            "W" -> intervalColors[1]
            "W+h" -> intervalColors[2]
            else -> intervalColors[2]
        }
    }
}